### ISSUE TITLE
flutter analyze should handle analyzer plugins

### ISSUE THREAD

**Issue by a14n** [flutter/flutter#28327]
I created an [analyzer_plugin to enforce Flutter Style Guide](https://github.com/a14n/flutter_style_guide_analyzer_plugin) and [registered it in flutter's packages](https://github.com/a14n/flutter/commit/558ed5910cfe32b2c2457153794a9926dcd77a89).

VSCode displays correctly the errors triggered by my plugin but `flutter analyze --flutter-repo` don't show any errors. (`flutter analyze --flutter-repo --watch` has the same behaviour)

Digging into the code `flutter analyze` stops once it receives an `{"isAnalyzing"::false}` event. But this event is sent before plugin responds.

I'm not sure what's the best way to fix that. Advices whould be very welcome.

/cc @devoncarew @scheglov @bwilkerson

**Comment by bwilkerson** [comment]
What signal is `flutter analyze` trying to use `server.status` for? I ask because it likely doesn't match the intent of the notification, which is merely to indicate to the client when the client might wish to display an indication that analysis server is actively analyzing files. It does _not_ mean that server won't send more analysis results, and, to the case in point, it was designed before there were plugins and doesn't indicate anything about the state of the plugins. (There isn't even a way, currently, for plugins to communicate that information to server.)

If `flutter analyze` needs a signal of some kind, then we can discuss adding one that has the right semantics, but I suspect that `server.status` is not the right signal for it to be using.

**Comment by devoncarew** [comment]
It is using `server.status` as an indication that analysis is done. It's definitely conflated with 'IDE busy' status.

Sounds like we may need some other way to indicate that the analysis results are up-to-date and available.

@MichaelRFairhurst ran into a similar situation recently with the angulardart analysis plugin. For some projects, it was taking significantly longer to return results than regular dart analysis, but IntelliJ (via the server.status notification) was showing any indication that processing was occurring. Perhaps extending server.status to include notifications about plugin activity would solve both issues?

**Comment by devoncarew** [comment]
@a14n, I see that you're using `import 'package:analyzer/src/dart/analysis/driver.dart';` in your plugin, and I assume are using resolved ASTs. Do note that we haven't well quantified the costs associated with running two copies of the analysis driver (also an issue for the angular dart plugin). Worst case, this could double processing time and memory usage; Flutter already watches the cold-start repo analysis time pretty closely.

**Comment by bwilkerson** [comment]
> Perhaps extending server.status to include notifications about plugin activity would solve both issues?

We could potentially do that, or we could roll them together (as in, report that we're analyzing any time server or any plugin is performing analysis). Either way, we'd need to add a way for plugins to report their status, and we'd be dependent on plugins doing so in order to use this signal.

The bigger problem I see is that this signal is being used in two different ways. When server is being used by an interactive client it's just an approximate indication of when results might be slow to be returned because server is still performing analysis that might be needed for those results. It isn't a guarantee that analysis is or is not being performed.

But now we want a guarantee for non-interactive clients that analysis is complete and all of the notifications that are ever going to be sent actually have been sent. That's a very different thing, and I don't think it's a good idea to conflate the two signals.

> Do note that we haven't well quantified the costs ...
> Worst case, this could double processing time and memory usage ...

You're right that we haven't quantified the costs. But I'm fairly confident that we're doubling both. As far as I know there isn't anything we can currently do to avoid that.

**Comment by a14n** [comment]
> Flutter already watches the cold-start repo analysis time pretty closely.

Every change to the codebase has a cost. I agree this plugin will add memory usage (and eventually processing time once `flutter analyze` correctly handle plugins) but it would highlight things that are today catched (or not) at review time (format, rules from [Flutter Style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)) and that are time consuming for the reviewer. IMHO it's worth a change.

FWIW I already use this plugin to catch [unhomogeneous formatting of parameter lists](https://github.com/flutter/flutter/pull/27261). Unfortunattely this formatting is not enforced today and @dnfield [requested a tool for that](https://github.com/flutter/flutter/pull/27261#issuecomment-464889250).

**Comment by devoncarew** [comment]
> Every change to the codebase has a cost. I agree this plugin will add memory usage (and eventually processing time once `flutter analyze` correctly handle plugins) but it would highlight things that are today catched (or not) at review time (format, rules from [Flutter Style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo)) and that are time consuming for the reviewer. IMHO it's worth a change.

I agree as well - I think it's worth the trade off for the additional time on the CIs. If the cold start time of `flutter analyze --flutter-repo` time does double however, it may be difficult to actually land the PR enabling the plugin.

**Comment by bwilkerson** [comment]
There is an alternative to keep in mind. If it's too expensive to have a separate plugin we could roll the functionality into server directly.

That said, the same overhead applies to the angular plugin, and as far as I know we have not heard any complaints from users about start up times there. I don't know that anyone has measured the cold start times with and without the plugin, though. (Might be an interesting data point.)

**Comment by dnfield** [comment]
What if we treated it was a separate process entirely under `flutter format`?

**Comment by bwilkerson** [comment]
Then we'd loose the IDE integration.

**Comment by dnfield** [comment]
Isn't there IDE integration for formatting the document?

**Comment by derolf** [comment]
I have a similar issue with my analyzer plugin. When using Visual Code, my reported errors show up, but if I run ‘flutter analyze’, they don’t appear at the console.

**Comment by zanderso** [comment]
@a14n I would love to turn on your formatting plugin for the flutter_tool package.

Getting this to work under `flutter analyze [--watch]` would be nice, but afaik, IDEs talk to the analysis server directly, not through the Flutter tool, so I don't think that would be a requirement. (It would be a nice to have, but not a blocker as far as I'm concerned.)

Having said that, I tried adding the plugin to the flutter_tool pubspec.yaml and referenced it from the flutter_tool analysis_options.yaml to try to repro the problem, but didn't see any messages from the plugin from my IDE, the analyzer CLI, or `flutter analyze` despite injecting some formatting issues. Does the plugin code need an update?

**Comment by a14n** [comment]
@zanderso the following config works for me with VSCode

Once the diff applied, restart VSCode. After some seconds you should see entries in _PROBLEMS_ panel.

```diff
diff --git a/packages/flutter_tools/analysis_options.yaml b/packages/flutter_tools/analysis_options.yaml
index 7d472a027..3ae40f87f 100644
--- a/packages/flutter_tools/analysis_options.yaml
+++ b/packages/flutter_tools/analysis_options.yaml
@@ -7,6 +7,8 @@ analyzer:
   strong-mode:
     implicit-casts: true
     implicit-dynamic: false
+  plugins:
+  - flutter_style_guide_analyzer_plugin
 
 linter:
   rules:
diff --git a/packages/flutter_tools/pubspec.yaml b/packages/flutter_tools/pubspec.yaml
index 516c5f23e..a9e6ea9fa 100644
--- a/packages/flutter_tools/pubspec.yaml
+++ b/packages/flutter_tools/pubspec.yaml
@@ -120,6 +120,10 @@ dependencies:
   watcher: 0.9.7+12 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
 
 dev_dependencies:
+  flutter_style_guide_analyzer_plugin:
+    git:
+      url: https://github.com/a14n/flutter_style_guide_analyzer_plugin.git
+      ref: b61d2e59391c63366831077481c51f73fceae9ba
   collection: 1.14.11
   mockito: 4.1.1
   file_testing: 2.1.0
```

**Comment by zanderso** [comment]
Thanks! I got your plugin working and was able to reproduce the issue somewhat. Results come back from the plugin during `flutter analyse --watch`, but they get queued up after an initial report, and are only printed to the console as part of a subsequent report.

Coming back around to this issue as stated, I would very much like for the tool to be able to support analyzer plugins, but because:
1. Plugins are loaded asynchronously (See [here](https://github.com/dart-lang/sdk/blob/863f7df7aad2d315d870b5c5ff5a64ec298c9280/pkg/analysis_server/lib/src/plugin/plugin_watcher.dart#L81), for example), and
2. There isn't a reliable signal that results are complete,

We probably won't be able to resolve this issue.

@devoncarew @bwilkerson if you don't already have issues filed for the analysis server for (1) and (2), I can go ahead and do that, otherwise if you could link them here that would be helpful.

**Comment by derolf** [comment]
Any update on the topic? What’s the current best way to add custom liner fuels?

**Comment by roman-petrov** [comment]
There is promising [dart_code_metrics](https://pub.dev/packages/dart_code_metrics) package that provides a number of helpful code diagnostics rules. It is configured as Flutter analyzer plugin through `analyzis_options.yaml` and works great in Visual Studio Code.  But currently I do not see how to use it in my CI pipeline via `flutter analyze` command.

It would be great if Flutter tooling will support extensibility as do many front end code quality tools like [ESLint](https://eslint.org/): it has great community and many plugins. This feature might help to improve overall `Dart` code quality including `Flutter` code base.

**Comment by zanderso** [comment]
The upstream Dart SDK issue to track is probably https://github.com/dart-lang/sdk/issues/38407

**Comment by bkonyi** [comment]
@bwilkerson is this still relevant?

**Comment by bwilkerson** [comment]
No, I believe this is now working.
